Fix cut/copy/paste/delete

chloebrett opened the pull request:
Massive PR, will break this up and it's a WIP for now, but it basically fixes the cut/copy/paste/delete editor functionality. It also fixes the issue of spaces disappearing when words are chopped and rearranged.

Video demo:

https://user-images.githubusercontent.com/6735055/181770373-b321364b-69ca-47d2-8953-15a0d620792b.mov

Video looks a bit random without keyboard shortcuts being shown, but basically I start by copying a phrase, then paste it in a few places, then do the same with cutting, then delete some words, then undo everything, then redo everything. All of this works as expected.

However... in doing so it breaks the video preview 😆 so I will have to look into that before finalising this. Hopefully that will be the easy part since I think it's just the transcript to cuts function we need to look into

Some other notes of interest
- Sets up a dummy transcription service to just return the sample transcription, pretty useful for those with M1's or who just don't want to wait a long time for transcriptions to finish
- Moves to UUIDs for word keys so that we don't get duplicate key errors (this causes some potential other problems but there should be workarounds for those)
- Re-creates spaces every time an edit is made so that there are always appropriate spaces between words
- Reworks the clipboard to store the actual words content rather than just indices, since the clipboard exists outside of time with respect to the undo stack, so rather than updating the clipboard every time we undo or redo, we can just store the actual content to be copied
- moves the live process function to the front end since it's only called from there, and calling it from the back end causes some weird polyfill stuff (might help with packaging, that could be part of the build error!)
- instead of testing liveProcess directly, we test it via the transcription reducer and I made a bunch of helper methods to simplify this process
- fixed a bunch of random off by one errors that were causing some of the UI weirdness with cut/copy/paste

Also, this PR builds on #132, so it will become marginally smaller once that's merged.

Resolves #141
Resolves #140
Resolves #139 
Resolves #138
Resolves #116
Resolves #113

Commit:
Many updates to edit/delete logic

Commit:
Fix liveProcess mutability bug

Commit:
Write transcription reducer tests and fix bugs identified

Commit:
Manipulating spaces in live transcription, but now tests are breaking as they don't expect spaces

Commit:
Fix tests for transcription reducer; remove live processing tests for now as live processing logic has been changed

Commit:
Move live process into front end; update some tests; fix undo paste logic

LiamTodd commented on src/main/editDelete/preProcess.ts:
```diff
@@ -19,89 +20,16 @@ const camelCase: MapCallback<SnakeCaseWord, PartialWord> = (word) => ({
- * @param words The list of words being reduced
- * @returns The updated transcript with a silence after word
- */
-const addSpaces: (totalDuration: number) => MapCallback<Word, Word[]> =
```
does the transcription object no longer use spaces to represent silences?

chloebrett replied:
No, this logic was moved to a shared file as it’s used both by live process (front end) and pre process (back end)

see processingShared.ts

However I think we may be able to simplify things and only add in spaces for (a) rendering the text and (b) video preview, which would make manipulating the transcript far easier

chloebrett replied:
Update: no longer using spaces

Commit:
Don't add spaces to the transcription itself; just add them during rendering

Commit:
Lots of changes; including but not limited to removing the notion of spaces as words and introducing before- and after- space buffers

Commit:
Fix edge case bug of pasting the same word multiple times in one paste

Commit:
Merge branch 'develop' into bugfix/cut-paste-deleted-words

chloebrett commented:
Done!!! 🎉

Here's a demo, showing the following:
- Moving around the timeline both when stopped and while video is playing
- Cutting, copying, pasting text
- Some edge cases
- Integration with undo

Uploading demo.mov…

As you can see, it all works more or less without a hitch (as far as I can tell from my own testing) - even edge cases like copying a word, pasting it next to itself, then copying both the original word and the pasted word, and pasting both of them somewhere else.

Now obviously this is an ungodly huge PR, unfortunately as it provides more or less integrated functionality that I can't really see being broken down into smaller parts, I think trying to break it into multiple PRs won't be a great idea. Instead, I propose:

1. First we get #132 reviewed, approved and merged, since this builds on that.
2. Then let's all sit down together (whoever is interested in this PR, I'm thinking @rileykeane, @aretle, @LiamTodd maybe?) and we'll go through it, you can ask questions and give feedback while I explain the changes I've made
3. As a side note, a significant portion of the line changes are from the test files, so that counts for something 🙂

Commit:
Move use dummy to config and disable it

Commit:
Merge branch 'develop' into bugfix/cut-paste-deleted-words

chloebrett reviewed (commented):
Don't need space at start or end of rendering

Tag Hugh

Dan: why can't we just use "current length of transcript plus one" as the key each time we paste / create new keys? -> I think there was an issue with this but I can't remember specifically what

Liam: what if we use "original index concatted with new index" as a key

hhen0005 commented:
Just doing some testing of the app and found an issue. 
If I cut and paste the first word into the middle then the video preview doesn't seem to get updated. 
The video preview will still the play the cut content (the clock and the transcript text preview do get updated though) and will not play the pasted content. 
Of course, once the paste point has been passed in the video playback, the bug is not noticeable.

chloebrett commented:
> Just doing some testing of the app and found an issue. If I cut and paste the first word into the middle then the video preview doesn't seem to get updated. The video preview will still the play the cut content (the clock and the transcript text preview do get updated though) and will not play the pasted content. Of course, once the paste point has been passed in the video playback, the bug is not noticeable. Let me know what you find please :)

Good find, will investigate. I think it's probably a bug in transcriptToCuts since the editor experience is unaffected

Commit:
Don't render spaces before and after transcription

Commit:
Update transcriptToCuts tests; fix bug in output start times calculator

Commit:
pasteCount -> pasteKey, simpler reducer algorithm

Commit:
Fix tests for transcription reducer to account for pasteKey logic change

chloebrett commented:
I've fixed all known issues:
- pasteCount is now pasteKey and the algorithm is O(n) -> doesn't need to take into account originalIndex, that was the simplification (thanks Dan)
- the bug that Hugh identified is fixed - it originated from the function that calculates output start times erroneously assuming the first word in the array would also be the first word in the original transcript
- Spaces are no longer rendered before and after the first/last word in the TranscriptBlock component

I think that addresses all issues raised, please re-review and approve if you are happy with the PR :)

rileykeane commented on src/renderer/store/transcription/reducer.ts:
```diff
@@ -26,65 +28,78 @@ const transcriptionReducer: Reducer<Transcription | null, Action<any>> = (
+    // Paste key must be unique for all pasted words - that is, no two pasted words should ever have the same paste key.
+    // We force this invariant by finding the highest paste key in the entire transcription to this point, and then
+    // adding n to it for the nth pasted word, for all words on the clipboard.
+    const highestExistingPasteKey = Math.max(
```
😍

danfeltham reviewed (approved):
Happy for this to be merged.

LiamTodd commented:
Looks good, thanks for explaining everything @chloebrett

Commit:
Merge branch 'develop' into bugfix/cut-paste-deleted-words
